core/drivers/imx_ele.c

```diff
@@ -0,0 +1,576 @@
+
+	return imx_ele_call(&msg);
+}
+
+/*
+ * Get the current state of the ELE TRNG
+ */
+static TEE_Result imx_ele_rng_get_trng_state(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rng_get_trng_state_msg_rsp {
+		uint32_t rsp_code;
+		uint8_t trng_state;
+		uint8_t csal_state;
+	} __packed rsp = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_BASELINE,
+		.header.size = 1,
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_TRNG_STATE,
+	};
+
+	res = imx_ele_call(&msg);
+	if (res)
+		return res;
+
+	memcpy(&rsp, msg.data.u8, sizeof(rsp));
+
+	if (rsp.trng_state != IMX_ELE_TRNG_STATUS_READY)
+		return TEE_ERROR_BUSY;
+	else
+		return TEE_SUCCESS;
+}
+
+/*
+ * Get random data from the EdgeLock Enclave
+ *
+ * @buffer RNG data output
+ * @size RNG data size
+ */
+static TEE_Result imx_ele_rng_get_random(uint8_t *buffer, size_t size)
+{
+	uint64_t timeout = timeout_init_us(10 * 1000);
+	paddr_t pa = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rng_get_random_cmd {
+		uint32_t addr_msb;
+		uint32_t addr_lsb;
+		uint32_t size;
+		uint32_t crc;
+	} cmd = { };
+	struct imx_mu_msg msg = {
+		.header.version = ELE_VERSION_HSM,
+		.header.size = SIZE_MSG_32(cmd),
+		.header.tag = ELE_REQUEST_TAG,
+		.header.command = ELE_CMD_RNG_GET,
+	};
+
+	assert(buffer && size);
+
+	if (cpu_mmu_enabled())
+		pa = virt_to_phys(buffer);
+	else
+		pa = (paddr_t)buffer;
+
+	reg_pair_from_64((uint64_t)pa, &cmd.addr_msb, &cmd.addr_lsb);
+	cmd.size = size;
+
+	/*
+	 * Check the current TRNG state of the ELE. The TRNG must be
+	 * started with a command earlier in the boot to allow the TRNG
+	 * to generate enough entropy.
+	 */
+	while (imx_ele_rng_get_trng_state() == TEE_ERROR_BUSY)
+		if (timeout_elapsed(timeout))
+			return TEE_ERROR_BAD_STATE;
+
+	memcpy(msg.data.u8, &cmd, sizeof(cmd));
+
+	update_crc(&msg);
+
+	if (cpu_mmu_enabled())
+		cache_operation(TEE_CACHEFLUSH, (void *)buffer, size);
+
+	res = imx_ele_call(&msg);
+
+	if (cpu_mmu_enabled())
+		cache_operation(TEE_CACHEINVALIDATE, (void *)buffer, size);
+
+	return res;
+}
+
+unsigned long plat_get_aslr_seed(void)
+{
+	unsigned long aslr __aligned_ele = 0;
+
+	if (imx_ele_rng_get_random((uint8_t *)&aslr, sizeof(aslr)))
+		panic("Fail to the seed the ASLR");
+
+	return aslr;
+}
+
+#define UID_SIZE (4 * sizeof(uint32_t))
+int tee_otp_get_die_id(uint8_t *buffer, size_t len)
+{
+	uint32_t session_handle = 0;
+	/*
+	 * The die ID must be cached because some board configuration prevents
+	 * the MU to be used by OPTEE at runtime.
+	 */
+	static struct session_get_device_info_rsp rsp;
+
+	if (rsp.rsp_code)
+		goto out;
+
+	if (imx_ele_session_open(&session_handle))
+		goto err;
+
+	if (imx_ele_session_get_device_info(session_handle, &rsp))
+		goto err;
+
+	if (imx_ele_session_close(session_handle))
+		goto err;
+
+out:
+	/*
+	 * In the device info array return by the ELE, the words 2, 3, 4 and 5
+	 * are the device UID.
+	 */
+	memcpy(buffer, rsp.chip_uid, MIN(UID_SIZE, len));
```

Comment:
suggestion: s/UID_SIZE/sizeof(rsp.chip_uid)/   (makes macro UID_SIZE useless)

Reply:
I take it